### PR TITLE
Bugfix: Bring back singleton option when creating a new `Session Bean` and interfaceless EJBs

### LINKED ISSUE #4830: Jakarta EE 9.1 - Impossible to create an EJB without an interface (using "New Session Bean")
### Apache NetBeans version

Apache NetBeans 16 release candidate

### What happened

Creation of a new session bean (EJB):
We must "choose" (it's the only option!) to create a local interface. It's not possible to create an EJB without an interface.

### How to reproduce

Right-clic on the project or a package and choose "New Session Bean".
You have to check the box "Local" for the interface. Else the EJB is not created.

### Did this work correctly in an earlier version?

Apache NetBeans 15

### Operating System

Windows 11 Family 64-bit, version 21H2

### JDK

11

### Apache NetBeans packaging

Apache NetBeans provided installer

### Anything else

_No response_

### Are you willing to submit a pull request?

No

### Code of Conduct

Yes

### LINKED ISSUE #4892: No more option for creating an EJB singleton in the wizard that creates an EJB
### Apache NetBeans version

Apache NetBeans 16 release candidate

### What happened

In a Jakarta EE Web application, the previous versions of NetBeans allowed to create an EJB singleton with New > Session Bean. It is no longer the case.
It would be nice to have it back.

### How to reproduce

Create a Maven Jakarta EE Web application.
Right click on the application and choose New > Session Bean. No radio button for Singleton.
I join 2 snapshots: with a previous version and with the current beta version.

current
previous

### Did this work correctly in an earlier version?

Apache NetBeans 15

### Operating System

Windows 11 Family 64-bit, version 21H2

### JDK

11

### Apache NetBeans packaging

Apache NetBeans provided installer

### Anything else

It is not a big problem but I don't understand why this option was removed.
I reported other errors in this wizard that creates an EJB and perhaps it is not difficult to add this option back in the same time.

### Are you willing to submit a pull request?

No

### Code of Conduct

Yes

### PR DESCRIPTION
**Pull request by asbachb**
The options shown when selecting `New > Session Bean` are based on the capabilities of the used enterprise project.

The check missed some Jakarta EE versions and so were disabled.

As EJB and CDI specs are backwards compatible the code was refactored so that only a minimal version and the profile type is checked to match that EJB or CDI version.

This fixes #4892 and fixes #4830

### CODE REVIEW

**Comment by asbachb** [comment]
Note: Recognized that these wrongly implemented EJB checks are causing - at least parts - of the problems of other broken Jakarta EE functionalities. e.g. part of problems of #4829 is that EJB check returns false for Jakarta EE.

**Comment by asbachb** [comment]
@matthiasblaesing  So my understanding was that these methods are used to determine if a specific feature is available in the used EJB version. From what I know that the enterprise specs should be backward compatible.

> 
> New features, enhancements or additions
> 
>     Moved to the jakarta.* namespace
> 
> Removals, deprecations or backwards incompatible changes
> 
>     Removal of methods relying on java.security.Identity
>     Removal of methods relying on JAX-RPC
>     Removal of deprecated EJBContext.getEnvironment() method
>     Removal of Support for Distributed Interoperability
>     Mark optional EJB 2.x API Group
> 

(see https://jakarta.ee/specifications/enterprise-beans/4.0/)

Beside the change of namespace I don't see any indication why this shouldn't be handled as functional backward compatible.

Basically I think it would be a better approach to expect the specs to be backwards compatible and fix stuff breaking in case. The opposite would be extend the `if` statements on every new enterprise release which causes much more delay and problems from my point of view.

**Comment by asbachb** [comment]
Interestingly the spec mention the namespace change as "new feature" and not as "backwards incompatible" or breaking change.

I still think from a functional standpoint the question if 4.0 is compatible with 3.2 the answer is yes. From a technical standpoint it might be a "kind of".

The question about which namespace is used I'd consider as a technical one and the question if for example a remote ejb needs an interface a functional one.

Currently to answer both questions there's only one class. Maybe that's the reason for the argument?



25 Jul 2023 02:48:02 Matthias Bläsing ***@***.***>:

**Comment by asbachb** [comment]
From the Jakarta EE 9 spec, introducing EJB 4.0:

> Due to the migration from the javax namespace to the jakarta namespace, Jakarta EE 9 is not source-code compatible or binary compatible with previous releases. However, Jakarta EE 9 is behavior compatible with all jakarta APIs present sharing identical method signatures and behavior with their javax equivalents.
> 
> Applications migrating to Jakarta EE 9 from previous releases will require modification in the areas detailed in the sections below.

source https://jakarta.ee/specifications/platform/9/jakarta-platform-spec-9.html#a3901

**Comment by matthiasblaesing** [comment]
Looking at `J2eeProjectCapabilitiesTest` I think, that it validates my interpretation. If I remember correctly you guessed the right package name for session beans by examining the classpath, but consider, that you might not want to do classpath scaning, then you could query the EJB capability and decide based on that.

I will not merge as is, but I won't stand in the way if @jGauravGupta, @pepness or @juneau001 want to merge anyway. They worked on this before, so might feel different then me.

**Comment by asbachb** [comment]
@matthiasblaesing what do you think about splitting the question if the used ejb spec is technical compatible vs functional compatible?

Like `TechnicalEJBCompatibility` vs `FunctionalEJBCompatibility` where things could be handled differently.

Big picture: I want to lower the maintenance efforts needs to be done by a maintainer / contributor. Fact is that some enterprise features are broken (like the missing button) based on the fact that a new method was introduced to `J2eeProjectCapabilities`, but the calls to the `isEjbXX` was not analyzed. So NetBeans interpreted this that EJB 4 does not provide the same functionality as EJB 3.1. When we'd assume that a new EJB release is backward compatible - like most of the time - we would not run in that situation. With EJB 4 it might create wrong namespaces, but at least the features are not completely absent / broken. Don't know if `J2eeProjectCapabilities` is the right place to do that. But from my point of view assuming that a new release is backward compatible is much more likely than assuming a maintainer / contributor will add new Jakarta EE releases on all necessary places (as this is the reason why the features are fully broken).

**Comment by matthiasblaesing** [comment]
@asbachb I have mixed feelings here, but I doubt, that we see the same screw up again. I assume, that the next Jakarta EE version will retain its backwards compatibilty. The massive breaking going from JavaEE to JakartaEE was expected and if it is repeated, I assume, that the platform will die. So I assume that the question `isEjb40Supported` and `isEjb40LiteSupported` both can be answered positively for the next specification versions, because it is true. If they again break namespaces or packages, well, then yes, it will break again and IMHO validly so.

I'm not the one that did the work the last time round, so the opinions of @pepness, @juneau001, @jGauravGupta  and @OndroMih would be helpful.

**Comment by OndroMih** [comment]
Hi, my gut feeling is as long as the component version is supported by the latest Jakarta EE, it’s OK to assume that it will continue to be supported for future versions. In the worst case, older versions of Netbeans will generate code that will need to be fixed by the user and newer versions of NB will fix the behavior to return false for newer Jakarta EE versions that break the API. IMHO it’s better than not being able to generate any code. E.g. isEjb40Supported() would be OK as in this PR.

But for older versions, e.g. is Ejb31Supported, they should only return true for Jakarta EE versions that don’t break the API. In this case, Jakarta EE 9 is no linger compatible with EJB 3.1, so isEjb31Supported() would be true only for JavaEE 6-8 and Jakarta EE 8.

This way, we’d only need to fix Netbeans in case there’s another breaking change in Jakarta EE.

**Comment by asbachb** [comment]
> But for older versions, e.g. is Ejb31Supported, they should only return true for Jakarta EE versions that don’t break the API.

I wonder if that's the assumption if `supported` is the right phrase. Aren't we talking with that assumption more about `compatibility` than `support`?

**Comment by OndroMih** [comment]
These `isEjbXXSupported` methods are used to offer generating code. The code generated for different versions of EJB may be different. But I assume that it shouldn't break the build. So, if the project supports EJB 3.2, it's OK to generate code for EJB 3.0, 3.1 or 3.2. Code generated for EJB 4.0 wouldn't compile because it uses jakarta package prefix, so it's not good to offer it for projects that only support EJB 3.2 and not 4.0.

The other way around should also apply - options for generating EJB 3.2 code shouldn't be available for EJB 4.0 projects because then the generated code wouldn't compile.

If we want to bring back the option to generate Singleton EJBs for EJB 4.0, then we need to fix the generator to support EJB 4.0 and `jakarta` prefix. This is what I think has been done for Stateless and Stateful, but not for Singleton.

What I suggest to do with `isEjbXXSupported` and similar methods, is that we allow one exception - if EJB 4.0 is the latest version of EJB, then Netbeans offers generating EJB 4.0 code also for any newer EJB version, e.g. EJB 5.0. If EJB 5.0 introduces a breaking change, then older versions of Netbeans would allow generating EJB 4.0 code with EJB 5.0, which could generate code that doesn't compile. But this would be treated as a bug and newer versions of Netbeans should fix that and change the `isEjb40Supported` method to not allow EJB 4.0 code generation for EJB 5.0+ projects.

**Comment by pepness** [comment]
I agree with @OndroMih. Jakarta EE 9/9.1/10 does not support EJB 3.2 and prior because it will not compile, even if the only change is the namespace, xml, and xds files.

**Comment by matthiasblaesing** [comment]
There was not update since the last milestone move. Given that the issue (singleton option for session beans) was fixed, this shifts focus to make transitions between specification versions easier. There is no consesus here to continue down this road, so I'm closing this.
